subsys/bluetooth/mesh/shell/shell_onoff_cli.c

```diff
@@ -0,0 +1,128 @@
+	}
+
+	struct bt_mesh_onoff_cli *cli = mod->user_data;
+
+	return bt_mesh_onoff_cli_get(cli, NULL, NULL);
+}
+
+static int onoff_set(size_t argc, char *argv[], bool acked)
+{
+	if (argc < 3) {
+		return -EINVAL;
+	}
+
+	bool on_off = (bool)strtol(argv[1], NULL, 0);
+	uint32_t time = (uint32_t)strtol(argv[2], NULL, 0);
+	uint32_t delay = (uint32_t)strtol(argv[3], NULL, 0);
+
+	if (!mod) {
+		mod = shell_model_first_get(BT_MESH_MODEL_ID_GEN_ONOFF_CLI);
+	}
+
+	struct bt_mesh_onoff_cli *cli = mod->user_data;
+	struct bt_mesh_model_transition trans = { .time = time, .delay = delay };
+	struct bt_mesh_onoff_set set = { .on_off = on_off, .transition = &trans };
+
+	if (acked) {
+		return bt_mesh_onoff_cli_set(cli, NULL, &set, NULL);
+	} else {
+		return bt_mesh_onoff_cli_set_unack(cli, NULL, &set);
+	}
+}
+
+static int shell_onoff_set(const struct shell *shell, size_t argc, char *argv[])
+{
+	return onoff_set(argc, argv, true);
+}
+
+static int shell_onoff_set_unack(const struct shell *shell, size_t argc, char *argv[])
+{
+	return onoff_set(argc, argv, false);
+}
+
+static int shell_instance_set(const struct shell *shell, size_t argc, char *argv[])
+{
+	if (argc < 1) {
+		return -EINVAL;
+	}
+
+	struct bt_mesh_model *mod_temp;
+	uint16_t addr = (uint16_t)strtol(argv[1], NULL, 0);
+
+	mod_temp = shell_model_get(BT_MESH_MODEL_ID_GEN_ONOFF_CLI, addr);
+
+	if (mod_temp) {
+		mod = mod_temp;
+	} else {
+		shell_error(shell, "Unable to find model instance for element address %d", addr);
+	}
+
+	return 0;
+}
+
+static int shell_instances_get(const struct shell *shell, size_t argc, char *argv[])
+{
+	uint8_t elem_cnt = bt_mesh_elem_count();
+	uint16_t mod_arr[elem_cnt];
+
+	memset(mod_arr, 0, sizeof(mod_arr));
+	shell_model_instances_get(BT_MESH_MODEL_ID_GEN_ONOFF_CLI, mod_arr, ARRAY_SIZE(mod_arr));
+
+	for (size_t i = 0; i < ARRAY_SIZE(mod_arr); i++) {
+		if (mod_arr[i]) {
+			shell_print(shell, "Onoff client instance found at element addr: %d",
+				    mod_arr[i]);
+		}
+	}
+
+	return 0;
+}
+
+SHELL_STATIC_SUBCMD_SET_CREATE(
+	onoff_cmds, SHELL_CMD_ARG(onoff_get, NULL, "OnOff get (No args)", shell_onoff_get, 1, 0),
+	SHELL_CMD_ARG(onoff_set, NULL, "OnOff set <onoff> <time> <delay>", shell_onoff_set, 4, 0),
+	SHELL_CMD_ARG(onoff_set_unack, NULL, "OnOff set unacknowledged <onoff> <time> <delay>",
+		      shell_onoff_set_unack, 4, 0),
+	SHELL_CMD_ARG(model_instance_set, NULL, "Set client instance <addr>", shell_instance_set, 2,
+		      0),
+	SHELL_CMD_ARG(model_instances_get, NULL, "Get client instances (No args)",
+		      shell_instances_get, 1, 0),
+	SHELL_SUBCMD_SET_END);
+
+static int cmds_temp(const struct shell *shell, size_t argc, char **argv)
+{
+	if (argc == 1) {
+		shell_help(shell);
+		/* shell returns 1 when help is printed */
+		return 1;
+	}
+
+	shell_error(shell, "%s unknown parameter: %s", argv[0], argv[1]);
+
+	return -EINVAL;
+}
+
+SHELL_CMD_ARG_REGISTER(mesh_onoff_cli, &onoff_cmds, "Onoff Cli commands",
```

Comment:
It would be really cool if we could get rid of `mesh-` prefix in commands here so that all commands are magically become subcommands of the `mesh` command, defined in `mesh/shell.c` in the stack. But of course this could be investigated later.

Reply:
I have investigated this a bit, but I was not successful in adding subsets of shell commands to an existing level0 command, e.g. `uart:~$ mesh onoff ...`. Then Trond told me that it was not a "must have" feature, so I abandoned it. It might be possible, but I think it requires a clever work-around to make it possible.